### PR TITLE
add Swap to @airswap/libraries

### PR DESCRIPTION
**Pull request by dmosites**

### CODE REVIEW

**Review by piersss — APPROVED** [review]

**Review by piersss — CHANGES_REQUESTED** [review]
Let me test this first 😋

File: tools/libraries/src/Swap.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    if (chainId in swapDeploys) {
+      return swapDeploys[chainId]
+    }
+    throw new Error(`Swap contract not found for chainId ${chainId}`)
+  }
+
+  public async check(
+    order: Order,
+    senderWallet: string,
+    signer?: ethers.providers.JsonRpcSigner
```
Comment (piersss) [inline tools/libraries/src/Swap.ts:36]:
```suggestion
    signer: ethers.providers.JsonRpcSigner
```

Why make it optional?

File: tools/libraries/src/Swap.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      } else {
+        contract = contract.connect(signer)
+      }
+    }
+    const [errors, count] = await contract.check(order)
+    return checkResultToErrors(count, errors)
+  }
+
+  public async swap(
+    order: Order,
```
Comment (piersss) [inline tools/libraries/src/Swap.ts:51]:
```suggestion
    order: FullOrder,
```

File: tools/libraries/src/Swap.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  public static getAddress(chainId = chainIds.MAINNET): string {
+    if (chainId in swapDeploys) {
+      return swapDeploys[chainId]
+    }
+    throw new Error(`Swap contract not found for chainId ${chainId}`)
+  }
+
+  public async check(
+    order: Order,
```
[the remark this answers is not in the document]
Comment (dmosites) [reply tools/libraries/src/Swap.ts:34]: Actually just `Order` as `FullOrder` includes `Settlement` which isn't passed here.

File: tools/libraries/src/Swap.ts (hunk 2 above)
Comment (dmosites) [reply tools/libraries/src/Swap.ts:51]: Same as above re `Order` vs `FullOrder`

File: tools/libraries/src/Swap.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (piersss) [reply tools/libraries/src/Swap.ts:34]: aha!

File: tools/libraries/src/Swap.ts (hunk 1 above)
Comment (dmosites) [reply tools/libraries/src/Swap.ts:36]: If a signer is passed to the constructor it'll be used, otherwise one must be passed here.

**Comment by piersss** [comment]
Using `check` I get this error:

```
Unhandled Rejection (Error): missing revert data in call exception; Transaction reverted without a reason string [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (data="0x", transaction={"from":"0xf450adf1807660A83bB0A4CeaE8E820283370812","to":"0xCb3839A953533339273e251Dd179515b7bA9a309","data":"0x3ceb2fc30000000000000000000000000000000000000000000000000000018748d9749500000000000000000000000000000000000000000000000000000000642b47c700000000000000000000000000000000000000000000000000000000000000070000000000000000000000002de63f2d35943aa17aa835ea69fd3f768b9f6337000000000000000000000000d000f000aa1f8accbd5815056ea32a54777b2fc480ac58cd0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dc00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d000f000aa1f8accbd5815056ea32a54777b2fc436372b07000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002386f26fc1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001cc4e3115c0644363a8f533370b768bcbdadd6e1869b2786ed3b89b8469617db8f710c655d97cff85c49b00cbb5bae6c32847ee46e10105d41dfb518a45193dec3","accessList":null}, error={"code":-32000,"message":"execution reverted"}, code=CALL_EXCEPTION, version=providers/5.7.2)
```

File: tools/libraries/src/Swap.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import { Swap as SwapContract } from '@airswap/swap/typechain/contracts'
+import { ISwap } from '@airswap/swap/typechain/contracts/interfaces'
+import { Swap__factory } from '@airswap/swap/typechain/factories/contracts'
+import { checkResultToErrors } from '@airswap/utils'
+
+import * as swapDeploys from '@airswap/swap/deploys.js'
+
+export class Swap {
+  public contract: SwapContract
+  public chainId: number
```
Comment (piersss) [inline tools/libraries/src/Swap.ts:12]: chainId is not used

**Comment by piersss** [comment]
Using `swap` I get a metamask prompt and then a transfer.

But the transfer is wrong. It's from sender wallet to sender wallet.

https://goerli.etherscan.io/tx/0x94bb8c2312257ae4e5e9e2828928c6dcd7eaa88fb80257a430618f3bab7424a7

This is the order I used:

```
affiliateAmount: "0"
affiliateWallet: "0x0000000000000000000000000000000000000000"
chainId: 5
expiry: "1680641799"
nonce: "1680638199853"
protocolFee: "7"
r: "0xfea98342f198551f4ba7b8e1cf03c4ed4259da224e9668aa22ba9ab426ce91aa"
s: "0x2f175c03e87e81f9d5e91b7513c8f3d5ad0acef7060fe3dab91ca32f3ec9c3a3"

sender: 
amount: "10000000000000000"
id: "0"
kind: "0x36372b07"
token: "0xd000f000aa1f8accbd5815056ea32a54777b2fc4"
wallet: "0x0000000000000000000000000000000000000000"

signer: 
amount: "0"
id: "220"
kind: "0x80ac58cd"
token: "0xd000f000aa1f8accbd5815056ea32a54777b2fc4"
wallet: "0x2de63f2d35943aa17aa835ea69fd3f768b9f6337"

swapContract: "0xCb3839A953533339273e251Dd179515b7bA9a309"
v: "27"
```

Used `0xf450adf1807660A83bB0A4CeaE8E820283370812` as senderWallet

File: tools/libraries/src/Swap.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { Swap__factory } from '@airswap/swap/typechain/factories/contracts'
+import { checkResultToErrors } from '@airswap/utils'
+
+import * as swapDeploys from '@airswap/swap/deploys.js'
+
+export class Swap {
+  public contract: SwapContract
+  public address: string
+
+  public constructor(signer: ethers.Signer, address?: string) {
```
Comment (piersss) [inline tools/libraries/src/Swap.ts:14]: `address` can be undefined?

**Comment by piersss** [comment]
Now it fails silently.

My code:

```
import * as swapDeploys from '@airswap/swap/deploys'; // (4.0.2)
import { Swap } from './ordersHelpers';

function swap(
  provider: ethers.providers.Web3Provider,
  order: FullOrder,
): Promise<ContractTransaction> {
  return new Swap(
    provider.getSigner(),
    swapDeploys[provider.network.chainId]).swap(
    order,
    '0',
  );
}
```

Maybe something unrelated but maybe it's because I run `tools/libraries/src/Swap.ts` locally? I put it in a file in the airswap-marketplace.

I tried linking the repo in package.json like so: `"@airswap/libraries": "../airswap-protocols/tools/libraries",`

But then when I run `yarn compile` to get build folders Lerna fails:

```
    ✖    5/9 targets failed, including the following:
         - @airswap/maker-registry:compile
         - @airswap/swap-erc20:compile
         - @airswap/staking:compile
         - @airswap/indexer-registry:compile
         - @airswap/swap:compile
```

**Pull request closed without merging**